src/Psalm/Internal/Analyzer/Statements/Expression/AssignmentAnalyzer.php

```diff
@@ -981,44 +975,20 @@ public static function analyzeAssignmentRef(
         }
 
         $lhs_location = new CodeLocation($statements_analyzer->getSource(), $stmt->var);
-        $statements_analyzer->registerVariableAssignment(
-            $lhs_var_id,
-            $lhs_location
-        );
+        if (!$stmt->var instanceof ArrayDimFetch && !$stmt->var instanceof PropertyFetch) {
+            // If left-hand-side is an array offset or object property, usage is too difficult to track,
+            // so it's not registered as an unused variable (this mirrors behavior for non-references).
+            $statements_analyzer->registerVariableAssignment(
+                $lhs_var_id,
+                $lhs_location
+            );
+        }
```

Comment:
It turns out the bug this pull request fixes actually affected this method, causing everything after `$context->vars_in_scope[$lhs_var_id] = &$context->vars_in_scope[$rhs_var_id];` to be ignored.

After it was fixed, that same line was reported as an unused variable, so I finally gave up and just decided that putting a reference in an object or an array is a bad idea anyway, and it's not worth it to track usage in that case (this is basically increasing the scope of the change in #7684 from only ignoring unused references on `$this` to ignoring unused references on any object or array).